src/Microsoft.TemplateEngine.Edge/Installers/NuGet/NugetApiPackageManager.cs

```diff
@@ -173,54 +173,47 @@ private async Task<(PackageSource, IPackageSearchMetadata)> GetLatestVersionInte
 
-            ConcurrentDictionary<PackageSource, IEnumerable<IPackageSearchMetadata>> searchResults = new ConcurrentDictionary<PackageSource, IEnumerable<IPackageSearchMetadata>>();
-            await Task.WhenAll(packageSources.Select(async source =>
-                {
-                    IEnumerable<IPackageSearchMetadata> foundPackages = await GetPackageMetadataAsync(source, packageIdentifier, includePrerelease: true, cancellationToken).ConfigureAwait(false);
-                    if (foundPackages == null)
-                    {
-                        return;
-                    }
-                    searchResults[source] = foundPackages;
-                })).ConfigureAwait(false);
-
-            if (!searchResults.Any())
+            if (!foundPackagesBySource.Any())
             {
                 throw new InvalidNuGetSourceException("Failed to load NuGet sources", packageSources.Select(source => source.Source));
             }
 
-            var accumulativeSearchResults = searchResults
-                .SelectMany(result => result.Value.Select(p => new { source = result.Key, package = p }));
+            var accumulativeSearchResults = foundPackagesBySource
+                .SelectMany(result => result.foundPackages.Select(package => (result.source, package)));
 
             if (!accumulativeSearchResults.Any())
             {
                 _nugetLogger.LogWarning($"{packageIdentifier} is not found in NuGet feeds {string.Join(", ", packageSources.Select(source => source.Source))}.");
                 throw new PackageNotFoundException(packageIdentifier, packageSources.Select(source => source.Source));
             }
 
-            var latestVersion = accumulativeSearchResults.Aggregate(
+            (PackageSource, IPackageSearchMetadata) latestVersion = accumulativeSearchResults.Aggregate(
                 (max, current) =>
                 {
-                    if (max == null) return current;
+                    if (max == default) return current;
                     return current.package.Identity.Version > max.package.Identity.Version ? current : max;
                 });
 
-            var latestStableVersion = accumulativeSearchResults.Aggregate(
+            (PackageSource, IPackageSearchMetadata) latestStableVersion = accumulativeSearchResults.Aggregate(
                 (max, current) =>
                 {
                     if (current.package.Identity.Version.IsPrerelease) return max;
-                    if (max == null) return current;
+                    if (max == default) return current;
                     return current.package.Identity.Version > max.package.Identity.Version ? current : max;
                 });
 
-            if (latestStableVersion != null && !includePreview)
+            if (latestStableVersion != default && !includePreview)
             {
-                return (latestStableVersion.source, latestStableVersion.package);
+                return latestStableVersion;
```

Comment:
If no stable version is found, we are directly returning a preview version if found. However, if `includePreview` is false, we should probably return `null`.

Reply:
this is intentionally to allow to install latest preview if no stable.

Reply:
I see. I believe the way nuget references work in dotnet today, unless you specifically asked for a preview, you will only be given a stable version or a "package not found" error. Wouldn't it be better to have the same experience across all dotnet? I will approve and leave this up to you.